### DISCUSSION TITLE
dagger functions/call/etc should fail if none present

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12374] [Help]
@kpenfound noticed this fun one while migrating to new the `source` fields:

```
❯ dagger init --sdk=go --name=playground --source .
Initialized module playground in .
❯ ls
dagger.gen.go  dagger.json  go.mod  go.sum  main.go  querybuilder
❯ rm main.go
❯ dagger functions
Name             Description
container-echo   example usage: "dagger call container-echo --string-arg yo stdout"
grep-dir         example usage: "dagger call grep-dir --directory-arg . --pattern GrepDir"
```

Even though there's *no* code in `main.go`, we still get some default generated functions - I assume this is due to changes in the auto-generation code, to always apply.

**Comment by jedevc** [comment]
cc @sipsma

**Comment by sipsma** [comment]
I'll confirm this is the case to verify there's no deeper issue here, but yeah I believe this is happening because the engine gives the module's context to the go SDK, it sees no `main.go` and fills in the default one, then the engine asks that what objects+functions it has and this is the result.

I think a lawyer could probably make a compelling case that this is *almost* correct behavior 😄 But it is undeniably weird and surprising.

I'd need to think about what the fix would be too. Right now the engine just always runs codegen on modules when loaded (relying on cache to save work), which keeps things simple and consistent. May need to add some kludge somewhere to conditionally not do that? But then that wouldn't handle the case where the dev hasn't run `dagger develop` (formerly `dagger sync`)... Unsure.

**Comment by helderco** [comment]
I think it's doable, see:
- https://github.com/dagger/dagger/issues/7034

Basically, always codegen, but only add from template when using `--sdk` (init/develop). There's a prerequisite that we need to flag when that happens to runtime modules.
